cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordPutTransformer.java

```diff
@@ -102,36 +117,54 @@ private void setField(Put put, Schema.Field field, Object val) {
     return type;
   }
 
-  private Put createPut(StructuredRecord record, Schema recordSchema) {
-    Schema.Field keyField = recordSchema.getField(rowField);
-    Preconditions.checkArgument(keyField != null, "Could not find key field in record.");
-    Object val = record.get(keyField.getName());
+  private Put createPut(StructuredRecord record, Schema.Field keyField) {
+    String keyFieldName = keyField.getName();
+    Object val = record.get(keyFieldName);
     Preconditions.checkArgument(val != null, "Row key cannot be null.");
 
     Schema.Type keyType = validateAndGetType(keyField);
     switch (keyType) {
       case BOOLEAN:
-        return new Put(Bytes.toBytes((Boolean) record.get(rowField)));
+        return new Put(Bytes.toBytes((Boolean) record.get(keyFieldName)));
       case INT:
-        return new Put(Bytes.toBytes((Integer) record.get(rowField)));
+        return new Put(Bytes.toBytes((Integer) record.get(keyFieldName)));
       case LONG:
-        return new Put(Bytes.toBytes((Long) record.get(rowField)));
+        return new Put(Bytes.toBytes((Long) record.get(keyFieldName)));
       case FLOAT:
-        return new Put(Bytes.toBytes((Float) record.get(rowField)));
+        return new Put(Bytes.toBytes((Float) record.get(keyFieldName)));
       case DOUBLE:
-        return new Put(Bytes.toBytes((Double) record.get(rowField)));
+        return new Put(Bytes.toBytes((Double) record.get(keyFieldName)));
       case BYTES:
-        Object bytes = record.get(rowField);
+        Object bytes = record.get(keyFieldName);
         if (bytes instanceof ByteBuffer) {
           return new Put(Bytes.toBytes((ByteBuffer) bytes));
         } else {
           return new Put((byte[]) bytes);
         }
       case STRING:
-        return new Put(Bytes.toBytes((String) record.get(rowField)));
+        return new Put(Bytes.toBytes((String) record.get(keyFieldName)));
       default:
         throw new IllegalArgumentException("Row key is of unsupported type " + keyType);
     }
   }
 
+  @Nullable
+  private Schema.Field getKeyField(Schema recordSchema) {
+    Schema.Field field = recordSchema.getField(rowField);
+    if (field == null && rowFieldCaseInsensitive) {
+      Iterable<Schema.Field> filtered = Iterables.filter(recordSchema.getFields(), new Predicate<Schema.Field>() {
+        @Override
+        public boolean apply(Schema.Field input) {
+          return input.getName().equalsIgnoreCase(rowField);
+        }
+      });
+      if (!Iterables.isEmpty(filtered)) {
+        Preconditions.checkArgument(Iterables.size(filtered) == 1, "Cannot have multiple fields in the schema that " +
+                                      "match %s in a case-insensitive manner when %s is enabled. Found %s.", rowField,
+                                    Properties.Table.CASE_INSENSITIVE_ROW_FIELD, Iterables.toString(filtered));
+        field = filtered.iterator().next();
```

Comment:
you could at this point set rowField to whatever you found.  Even if for some weird reason you get another record with a different casing, the search logic will still work to find it again.